### DISCUSSION TITLE
OAuth on vercel not working

### DISCUSSION THREAD

**Discussion by sebsch1337** [neuefische/questions#312] [Web] [answered]
Hello,

i implemented logins for Google, Facebook and GitHub. All of them work well on localhost. When i deploy them, i get the following messages:
Facebook:
![image](https://user-images.githubusercontent.com/75478893/200889796-347f4f51-0cef-4dc0-8b16-0a928b20923a.png)

Github & Google
![image](https://user-images.githubusercontent.com/75478893/200891095-d1c0a250-67e5-49b8-b478-a57adddcddac.png)

I already added the named domains in app domains:
![image](https://user-images.githubusercontent.com/75478893/200889954-5f08e4cf-55c5-405d-b3f6-9b4a248fddb2.png)

Also on Github:
![image](https://user-images.githubusercontent.com/75478893/200890451-99ba5205-8804-4689-99c2-1f06b7f31da1.png)

Vercel got all my secret for the new app
![image](https://user-images.githubusercontent.com/75478893/200890797-305dd09f-74f3-4920-b283-1b1ceeea808f.png)

Since all of these worked well, i assume theres a different problem on vercel.

PR: [https://github.com/sebsch1337/capstone-project/pull/50](https://github.com/sebsch1337/capstone-project/pull/50)
Link: [https://capstone-project-git-feat-user-login-sebsch1337.vercel.app/](https://capstone-project-git-feat-user-login-sebsch1337.vercel.app/)

What did i do wrong?

**Comment by pixelass** [comment]
Auth via Nextauth does not work on preview deployments. This can only be used on prod and local.

(for google this is possible but you have to add a new url for every push since vercel uses the deployment url which contains a new hash for every push)

GitHub: Needs a separate GitHub App per Environment
Google: Several subdomains can be added for one Google app
Facebook: Who cares about Facebook.

To make this possible you would have to provide your own auth handler (out of scope for capstone)

Here's an issue discussing this case: https://github.com/nextauthjs/next-auth/issues/497

Feel free to read, but I strongly advise you to not try to implement it.

**Reply by sebsch1337** [reply]
Actually it works in preview environment - i did it the same way as my colleague Mengzhen:
https://github.com/Mengzhen1992/capstone-project/pull/26

I added a new app for the preview and new secrets for it in vercel. (view the screenshots)

**Reply by pixelass** [reply]
Yes, but that means that you have to do it for every push/deployment.

**Reply by sebsch1337** [reply]
Unlike her deployment, I can't get any vercel deployment to run properly. So when i skip the preview and merge it into production, there will be the same bug. There must be something wrong in my code, the wrong secret definitions, or wrong settings for github / google / fb. But i cant find the problem, since theres not much to configure.

**Comment by FreytagDom** [comment]
Hello everyone, here is a supplement regarding NextAuth and the preview deployments that Vercel creates for our pull requests: Since the OAuth providers (Github, Google, etc.) do not work in these environments, you can configure a dummy login with the credentials provider especially for these deployments (e.g. with username="test" and password="test"). This way you can also test the features for logged-in users in the preview deployment.

**Reply by sebsch1337** [reply]
Thanks for the info. If i still have time, i'll consider a test login :)

**Reply by pixelass** [reply]
Here's an example of a credential login with credentials that uses GitHub secrets:

https://github.com/pixelass/pwa-template/blob/main/src/pages/api/auth/%5B...nextauth%5D.ts#L16-L46

**Reply by sebsch1337** [reply]
This looks pretty good, the comments are very helpful! Do you mean with "GitHub secrets" the matching credentials saved in env variables?

**Reply by pixelass** [reply]
You would store those in vercel (GitHub would only be needed if you want to use it for tests or other GitHub workflows)

**Comment by sebsch1337** [comment, marked as answer]
I made it!! So here is what i did.

### Keep an eye on the environment variable names
I got the wrong variable names from a tutorial (e.g. GOOGLE_ID instead of GOOGLE_CLIENT_ID). I'm not 100% sure if this was the mistake, but its not a big deal to make sure, that the provider env names match the ones described in the [docu](https://next-auth.js.org/providers/google)

### Add the env variables to vercel
It's important to add the variables from .env.local also in vercel. For every provider.

### Remove app access from your account
Always remove the app access from your accounts (e.g. on [permissions page](https://myaccount.google.com/permissions) of your google account) after you try to use a different domain. I guess the auth cookie is fixed to a domain, that will cause the error when trying to use that auth cookie on a different domain branch. I removed access and i was able to login again.

### Add the link to the provider services
Add the vercel link of your preview version (created when creating / pushing into your PR) to your auth service providers like google and keep them up-to-date when you push another commit.

Thanks for your support! :)
